src/hotspot/share/gc/g1/g1CodeRootSet.cpp

```diff
@@ -28,89 +28,280 @@
 #include "code/nmethod.hpp"
 #include "gc/g1/g1CodeRootSet.hpp"
 #include "gc/g1/heapRegion.hpp"
+#include "memory/allocation.hpp"
 #include "oops/oop.inline.hpp"
+#include "runtime/atomic.hpp"
+#include "utilities/concurrentHashTable.inline.hpp"
+#include "utilities/concurrentHashTableTasks.inline.hpp"
 
-void G1CodeRootSet::add(nmethod* nm) {
-  assert(_is_iterating == false, "should not mutate while iterating the table");
-  bool added = false;
-  if (_table == nullptr) {
-    _table = new (mtGC) Table(SmallSize, LargeSize);
+struct G1CodeRootSetHashTableValue {
+  nmethod* _nmethod;
+
+  G1CodeRootSetHashTableValue(nmethod* nmethod) : _nmethod(nmethod) { }
+};
+
+class G1CodeRootSetHashTableConfig : public StackObj {
+public:
+  using Value = G1CodeRootSetHashTableValue;
+
+  static uintx get_hash(Value const& value, bool* is_dead);
+
+  static void* allocate_node(void* context, size_t size, Value const& value) {
+    return AllocateHeap(size, mtGC);
+  }
+
+  static void free_node(void* context, void* memory, Value const& value) {
+    FreeHeap(memory);
+  }
+};
+
+// Storage container for the code root set.
+class G1CodeRootSetHashTable : public CHeapObj<mtGC> {
+  using HashTable = ConcurrentHashTable<G1CodeRootSetHashTableConfig, mtGC>;
+  using HashTableScanTask = HashTable::ScanTask;
+
+  // Default (log2) number of buckets; small since typically we do not expect many
+  // entries.
+  static const size_t Log2DefaultNumBuckets = 2;
+  static const uint BucketClaimSize = 16;
+
+  HashTable _table;
+  HashTableScanTask _table_scanner;
+
+  size_t volatile _num_entries;
+
+  bool is_empty() const { return number_of_entries() == 0; }
+
+  class HashTableLookUp : public StackObj {
+    nmethod* _nmethod;
+
+  public:
+    explicit HashTableLookUp(nmethod* nmethod) : _nmethod(nmethod) { }
+    uintx get_hash() const;
+    bool equals(G1CodeRootSetHashTableValue* value);
+    bool is_dead(G1CodeRootSetHashTableValue* value) const { return false; }
```

Comment:
I wonder if `(...)` works, since the arg in unused. (Inspired by `struct NOP` in conc-hashtable.)

Reply:
I meant `bool is_dead(...) const { return false; }`.